ocaml/xapi/xapi_vgpu_type.ml

```diff
@@ -533,6 +546,139 @@ module Intel = struct
     List.map (find_or_create ~__context) types
 end
 
+module MxGPU = struct
+  let amd_vendor_id = 0x1002
+
+  let ( *** ) = Int64.mul
+  let ( /// ) = Int64.div
+  let ( +++ ) = Int64.add
+  let ( --- ) = Int64.sub
+  let mib x = List.fold_left Int64.mul x [1024L; 1024L]
+
+  type vgpu_conf = {
+    (* The identifier has fields for framebuffer size and scheduling slice. *)
+    identifier : Identifier.mxgpu_id;
+    model_name : string;
+    share_of_pgpu : int64;
+  }
+
+  let read_whitelist_line ~line =
+    try
+      Some (Scanf.sscanf
+              line
+              "%04x name='%s@' framebuffer_sz=%Ld sched=%d share_of_pgpu=%Ld"
+              (fun pdev_id (* e.g. "FirePro S7150" has 6929 (PF), 692f (VF) *)
+                model_name (* e.g. PF "FirePro S7150" or VF "FirePro S7150V" *)
+                framebufferbytes
+                sched
+                share_of_pgpu ->
+                {
+                  identifier = Identifier.({
+                      pdev_id;
+                      sched;
+                      framebufferbytes;
+                    });
+                  model_name;
+                  share_of_pgpu;
+                }))
+    with e-> begin
+        error "Failed to read whitelist line: '%s' %s"
+          line (Printexc.to_string e);
+        None
+      end
+
+  let read_whitelist ~whitelist ~device_id =
+    if Sys.file_exists whitelist then begin
+      Stdext.Unixext.file_lines_fold
+        Identifier.(fun acc line ->
+            match read_whitelist_line ~line with
+            | Some conf when conf.identifier.pdev_id = device_id -> conf :: acc
+            | _ -> acc)
+        []
+        whitelist
+    end else []
+
+  let make_vgpu_types ~__context ~pci ~whitelist =
+    let open Xenops_interface.Pci in
+    let device_id =
+      Db.PCI.get_device_id ~__context ~self:pci
+      |> Xapi_pci.int_of_id
+    in
+    let address =
+      Db.PCI.get_pci_id ~__context ~self:pci
+      |> address_of_string
+    in
+    let whitelist = read_whitelist ~whitelist ~device_id in
+    let vendor_name, device =
+      Pci.(with_access (fun access ->
+          let vendor_name = lookup_vendor_name access amd_vendor_id in
+          let device =
+            List.find
+              (fun device ->
+                 (device.Pci_dev.domain = address.domain) &&
+                 (device.Pci_dev.bus = address.bus) &&
+                 (device.Pci_dev.dev = address.dev) &&
+                 (device.Pci_dev.func = address.fn))
+              (get_devices access)
+          in
+          vendor_name, device))
+    in
+    let size_of_pgpu = (* counterpart of bar_size in gvt-g *)
+      Constants.pgpu_default_size
+    in
+    let rec collect acc = function
+      | []             -> acc
+      | Some v :: tail -> collect (v :: acc) tail
+      | None   :: tail -> collect acc tail
+    in
+    whitelist
+    |> List.map
+      Identifier.(fun conf ->
+          let vgpus_per_pgpu =
```

Comment:
Again, this is to stay as close as possible to the form of the `Intel` module, for ease of future refactoring where I imagine we will have a function like `get_vgpus_per_pgpu` in the modules we pass to the functor.